Adds search index creation and search interface

p-galligan opened the pull request:
Adds a search interface and search index creation.

I think the styling on the results could use some work, but that is definitely not my strong suit right now.

You'll notice that I've enabled targeted searching here, but to do that I've had to do some variable setting based on the `searchType` because of a mismatch in the field names. I'm not sure what the best approach is here. I _could_ just change the names of the fields to "author" and "subject" in the `search_data.json` file here instead, but I'm not sure what the best approach is.

Commit:
adding basic lunr

Adding basic lunr.js.min support.

Cribbing `create-index` and `lunr-feed` from virtual vault.

Commit:
adding basic lunr

Adding basic lunr.js.min support.

Cribbing `create-index` and `lunr-feed` from virtual vault.

Commit:
add to gitignore

Add `node_modules` to gitignore

Commit:
add to gitignore

Add `node_modules` to gitignore

Commit:
fix gitignore

Fix directory path for node_modules

Commit:
fix gitignore

Fix directory path for node_modules

Commit:
fix directory path

Fixes directory path for search index build

Commit:
fix directory path

Fixes directory path for search index build

Commit:
Merge branch 'issue-12' of https://github.com/RockefellerArchiveCenter/library_discovery into issue-12

Commit:
Add search page

Add default search page to display search results.

Commit:
fix url in search_data

fix url syntax in `search_data.json`

Commit:
include ref in search_data

Includes `ref` in `search_data` for easier referencing in the search index.

Commit:
include searchType in results

Includes the searchType in the results, but doesn't target yet.

Commit:
check for search type

Checks length of `searchType` before including it in the results.

Commit:
tweak search results language

just small tweak to search results language

Commit:
implement targeted search

Adds targeted search functionality

Commit:
update cibuild

update cibuild to include index creation

Commit:
forgot to install lunr

Installs lunr in the js directory

Commit:
trying npm ci

Using npm ci for lunr

Commit:
update travis

Updates travis to install lunr before bundle

helrond commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ _site
 .sass-cache
 .jekyll-cache
 .jekyll-metadata
+js/node_modules/
```
I think you might have run `npm install` from within the `js` directory rather than the repository root, which is the general practice. I'd suggest gitignoring `node-modules/` and running `npm install` from the root.

helrond commented on _includes/scripts.html:
```diff
@@ -1,6 +1,10 @@
 
+<!-- Lunr.js for search -->
+<script type="text/javascript" src="//unpkg.com/lunr/lunr.js"></script>
+<script defer type="text/javascript" src="{{ site.url }}{{ site.baseurl }}/js/lunr-feed.js?3"></script>
```
You can remove the query parameter here - that's used to force browsers to fetch the script if you've updated it.

helrond commented on js/create-index.js:
```diff
@@ -0,0 +1,33 @@
+buildIndex();
+
+function buildIndex() {
+  fs.readFile('_site/search_data.json', 'utf8', function(err, data) {
```
I think you'll want to use some method that's a little more robust in terms of handling relative paths and being run from multiple places. Look at `__filename` and `__dirname`: https://stackoverflow.com/questions/3133243/how-do-i-get-the-path-to-the-current-script-with-node-js

p-galligan replied:
Not sure if my fix is what you were getting at, but I think it handles the situation.

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,68 @@
+      $.getJSON("search_data.json", function(documents){
+        for (r in results) {  // Iterate over the results
+          let item = documents[results[r].ref];
+          appendString += `<li><p class="lead mb-1"><a href="${item.url}">${item.title}</a><small>${item.call_numbers}</small><p><small>${item.agents}</small></p></li>`;
```
Can you format this so it's a little easier to read? Like nested HTML?

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,68 @@
+        $('#results').append(appendString);
+      });
+    }
+    if (searchType.length) {
```
Ditto, some formatting here would help make this more legible.

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,68 @@
+
+  let searchTerm = getQueryVariable('query');
+  let searchType = getQueryVariable('type');
+  if (searchType === "author") {
+    var searchField = "agents";
+  } else if (searchType === "subject") {
+    var searchField = "subjects";
+  } else if (searchType === "title") {
+    var searchField = searchType;
+  } else {
+    var searchField = "";
+  }
```
Yeah, we could avoid this if we just indexed things to the correct field, i.e. if we changed what's happening in `search_data.json` so that `agents` are `authors`.

p-galligan replied:
I'm still getting the hang of JS, so I still had to set `var searchField = searchType;`, but I think it's still cleaner this way.

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,68 @@
+      let index = lunr.Index.load(data)
+
+      if (searchField.length){
+        var results = index.search(`${searchField}:${searchTerm}`); // Get lunr to perform a search
```
Noice.

helrond reviewed (commented):
This looks REALLY good - I left a few minor comments but I'd be happy with merging this and dealing with the styling in another PR.

Commit:
update gitignore and package

updates gitignore and package-lock to include lunr

Commit:
update package-lock

updates package-lock

Commit:
remove query parameter

Commit:
use __dirname

Implements __dirname for better handling of different paths

Commit:
update index feed and create-index

Updates so index fields match search types and updates for formatting.

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  let searchTerm = getQueryVariable('query');
+  let searchType = getQueryVariable('type');
+  var searchField = searchType;
```
Is there a reason you're not just setting directly to `searchField`?

p-galligan replied:
So, for some reason when I tried to do `var searchType = getQueryVariable('type');` it broke page display/indexing, even when changing what we're passing to the targeted search below.

I was completely unable to figure out how to do it in a single line and have it still work.

helrond replied:
Look at the error that's getting thrown in the console. That indicates to me that you're using the variable `searchType` elsewhere in the script (the error indicates line 26).

helrond commented on js/create-index.js:
```diff
@@ -0,0 +1,34 @@
+          }
+      })
+
+      fs.writeFile('_site/search_index.json', JSON.stringify(idx), (err) => {
```
Again, you need to handle paths better here....

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,85 @@
+    $('#results').empty().append('<img class="mx-auto d-block" src="/img/loading.gif" />')
+    $('#query').attr("value", searchTerm);
+
+    $.getJSON("search_index.json", function(data){
```
Path handling needs to be improved here. If you go to `search/?query=dynastic` as opposed to `search?query=dynastic` (no trailing slash) this search no longer works.

helrond commented on js/lunr-feed.js:
```diff
@@ -0,0 +1,85 @@
+    if (results.length) { // Are there any results?
+      var appendString = '<ul class="list--unstyled">'
+
+      $.getJSON("search_data.json", function(documents){
```
Path handling needs to be improved here.

helrond replied:
For this and the other instance below, you can probably just point at an absolute path (i.e. `/search_data.json`) rather than the relative path here.

Commit:
remove searchType

Changes `searchType` to `searchField` and removes any reference to searchType.

Also includes absolute path to `search_data.json`.

Commit:
add absolute path

add absolute path for `search_index.json`

p-galligan commented:
I think I've addressed these issues, but let me know if I had the wrong ideas/approach.
